### ISSUE TITLE
Re-implement the plugin using PSI to allow more features to be integrated.

### ISSUE THREAD

**Issue by fcaps** [templ-go/templ-jetbrains#8]
Autocomplete for HTML and other "language contributors" like htmx would be so nice to have.
[languageInjection](https://plugins.jetbrains.com/docs/intellij/language-injection.html#implementation) could be a solution for this, right?

**Comment by emerson-argueta** [comment]
I think adding something like what is mentioned [here](https://stackoverflow.com/questions/35950280/intellij-custom-language-combine-fragments-in-other-languages) could help.

**Comment by fcaps** [comment]
tl;dr
Not a pro, but most likely this will not be possible without large adjustments or tech burdens.

Templ is using a LSP that is responsible for autocomplete and other editor functions, and since jetbrains is using PSI to configure features like language injection, there is no "easy" way to make it happen via a config or similar afaik. 

This is also the reason why usages, auto-import, refactoring and other "editor-functions" in .templ files are not working as in  normal .go files. 

At this point I am unsure how and if something like this will ever happen.
To make the DX as good as it can get in Golang with all the "candy" the base-languages are already providing like find usages, autocomplete htmx, use other plugins etc. looks complex, but maybe it's just my skill issue.

PS: noticed there is a [syntax file](https://github.com/templ-go/templ-jetbrains/blob/main/grammars/Syntaxes/templ.tmLanguage.json)  for the TemplHighlightingLexer.kt , but unsure if this could be used to at least make html/js/htmx autocomplete work.

PPS: how does neovim handle this kind of issues? are they all raw dogging it without proper linting and autocomplete for the html part?

**Comment by alexisvisco** [comment]
Not using the templ lsp seems something that we should look for. 

Jetbrains team provides some mechanisms to use their advanced language features. Not using them is a non-sense.

Using lsp of templ was a workaround more than a solution in my opinion. 

Vue and JSX are example of what templ should look like as a dx point of view: you can use intellij features of html and js/ts, you just don't notice the difference between a javascript regular file and a jsx file, it's the same with html. Templ should be the same. 

Having said all that, the best thing would be to find someone with the knowledge to do this or help us do it.

**Comment by fcaps** [comment]
totally agree, it should feel like JSX.
@joerdav you got any opinion about re-writing it in PSI?

What is missing?
- Autocomplete go (the current ac is just not as good as the one in .go files)
- Autocomplete html (autoclosing tags etc.)
- Syntax highlighting (still not the same as in normal go files) 
- Lint html (show if a tag is not known "<bar></bar>")
- Code formatting html/js/css
- Refactor function names (also the templ components)
- find usages (right click on templ component)
- jump to .templ and not into the transpiled *_templ.go if clicked on a component 
- Autocomplete, lint, highlight etc. JS and css
- should work with other plugins like https://plugins.jetbrains.com/plugin/20588-htmx-support

The most annoying part imo is the missing autocomplete in all 3 languages (go, html, js)

Just started goland a couple days ago, but something feels odd with all the templating packages, the ones that are trying to change that are limited by golang itself, or don't integrate in the common IDEs. 

Judging from videos of ThePrimeagen (neovim) and Anthony GG (vscode), other IDEs also not working as expected.

**Comment by alexisvisco** [comment]
Maybe we could implement a plugin that just 2 things and iterate over : 

- a templ file is a go func
- templ identifier(params...) { HERE IS HTML } 

In that way we could move around and adding feature. Starting simple could be a way to overcome the complexity of the plugin sdk of intellij.

**Comment by joerdav** [comment]
So I think the action here is to create a new plugin that is re-implemented using PSI. This could be on a branch of this repo or in a new one for now so that we don't break the existing one.

With that said I think this ticket is now up for grabs (I'll rename it too)

**Comment by myrjola** [comment]
Opened a WIP PR with some promising results https://github.com/templ-go/templ-jetbrains/pull/17.

> Maybe we could implement a plugin that just 2 things and iterate over :
> 
> a templ file is a actually a go file
> templ identifier(params...) { HERE IS HTML }
> In that way we could move around and adding feature. Starting simple could be a way to overcome the complexity of the plugin sdk of intellij.

Unfortunately, I could not make it yet work with both Go and HTML following the [Custom templating language plugin tutorial](https://intellij-support.jetbrains.com/hc/en-us/community/posts/206765105-Tutorial-Custom-templating-language-plugin). We have to choose either language and add support for the other with language injection in a later change. I could use some help in getting the PR ready for merging. Especially from lexer/parser experts 😄

**Comment by alexisvisco** [comment]
Hello @myrjola What's the next step through the improvement of the plugin ? Maybe trying to enhance go completions ?

**Comment by joerdav** [comment]
I actually think we can close this now! Any other enhancements we can raise as separate trackable tickets.
